File: tests/wasm/test_reverse_rsi.js

import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import {
    loadTestData,
    assertArrayClose,
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS
} from './test_utils.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {

    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const importPath = process.platform === 'win32'
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);

    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }

    testData = loadTestData();
});

test('REVERSE_RSI partial params', () => {

    const close = new Float64Array(testData.close);

    const result = wasm.reverse_rsi_js(close, 14, 50.0);
    assert.strictEqual(result.length, close.length);
});

test('REVERSE_RSI accuracy', () => {


    const close = new Float64Array(testData.close);


    const rsiLength = 14;
    const rsiLevel = 50.0;

    const result = wasm.reverse_rsi_js(close, rsiLength, rsiLevel);


    assert.strictEqual(result.length, close.length);



    const expectedLast5 = [
        60124.655535277416, 60064.68013990046, 60001.56012990757, 59932.80583491417, 59877.248275277445
    ];

    const last5 = result.slice(-6, -1);
    for (let i = 0; i < 5; i++) {
        assertClose(
            last5[i],
            expectedLast5[i],
            1e-6,
            `REVERSE_RSI last 5 values mismatch at index ${i}`
        );
    }
});

test('REVERSE_RSI default params', () => {

    const close = new Float64Array(testData.close);


    const result = wasm.reverse_rsi_js(close, 14, 50.0);
    assert.strictEqual(result.length, close.length);
});

test('REVERSE_RSI zero period', () => {

    const input_data = new Float64Array([10.0, 20.0, 30.0]);

    assert.throws(() => {
        wasm.reverse_rsi_js(input_data, 0, 50.0);
    }, /Invalid period/, 'Should throw error for zero period');
});

test('REVERSE_RSI period exceeds length', () => {

    const data_small = new Float64Array([10.0, 20.0, 30.0]);

    assert.throws(() => {
        wasm.reverse_rsi_js(data_small, 10, 50.0);
    }, /Invalid period/, 'Should throw error when period exceeds data length');
});

test('REVERSE_RSI invalid level', () => {

    const input_data = new Float64Array([10.0, 20.0, 30.0, 40.0, 50.0, 10.0, 20.0, 30.0, 40.0, 50.0,
                                          10.0, 20.0, 30.0, 40.0, 50.0, 10.0, 20.0, 30.0, 40.0, 50.0,
                                          10.0, 20.0, 30.0, 40.0, 50.0]);


    assert.throws(() => {
        wasm.reverse_rsi_js(input_data, 14, 150.0);
    }, /Invalid RSI level/, 'Should throw error for RSI level > 100');


    assert.throws(() => {
        wasm.reverse_rsi_js(input_data, 14, -10.0);
    }, /Invalid RSI level/, 'Should throw error for negative RSI level');
});

test('REVERSE_RSI edge levels', () => {

    const input_data = new Float64Array(50);
    for (let i = 0; i < 50; i++) {
        input_data[i] = 10.0 + (i % 5) * 10.0;
    }


    const result01 = wasm.reverse_rsi_js(input_data, 14, 0.01);
    assert.strictEqual(result01.length, input_data.length);

    assert(!result01.every(isNaN), 'Should have some valid values for RSI level = 0.01');


    const result1 = wasm.reverse_rsi_js(input_data, 14, 1.0);
    assert.strictEqual(result1.length, input_data.length);

    assert(!result1.every(isNaN), 'Should have some valid values for RSI level = 1.0');


    const result99 = wasm.reverse_rsi_js(input_data, 14, 99.0);
    assert.strictEqual(result99.length, input_data.length);

    assert(!result99.every(isNaN), 'Should have some valid values for RSI level = 99.0');
});

test('REVERSE_RSI various levels', () => {

    const input_data = new Float64Array(50);
    for (let i = 0; i < 50; i++) {
        input_data[i] = 10.0 + i;
    }

    const levels = [20.0, 30.0, 50.0, 70.0, 80.0];
    const results = [];

    for (const level of levels) {
        const result = wasm.reverse_rsi_js(input_data, 14, level);
        assert.strictEqual(result.length, input_data.length);
        results.push(result);
    }


    for (let i = 0; i < results.length - 1; i++) {

        let foundDifference = false;
        for (let j = 0; j < results[i].length; j++) {
            if (!isNaN(results[i][j]) && !isNaN(results[i+1][j])) {
                if (Math.abs(results[i][j] - results[i+1][j]) > 1e-10) {
                    foundDifference = true;
                    break;
                }
            }
        }
        assert(foundDifference, `Different RSI levels (${levels[i]} vs ${levels[i+1]}) should produce different results`);
    }
});

test('REVERSE_RSI empty input', () => {

    const input_data = new Float64Array([]);

    assert.throws(() => {
        wasm.reverse_rsi_js(input_data, 14, 50.0);
    }, /Input data slice is empty/, 'Should throw error for empty input');
});

test('REVERSE_RSI all NaN', () => {

    const input_data = new Float64Array(30);
    input_data.fill(NaN);

    assert.throws(() => {
        wasm.reverse_rsi_js(input_data, 14, 50.0);
    }, /All values are NaN/, 'Should throw error for all NaN input');
});

test('REVERSE_RSI insufficient data', () => {



    const input_data = new Float64Array([1.0, 2.0, 3.0, 4.0, 5.0]);

    assert.throws(() => {
        wasm.reverse_rsi_js(input_data, 14, 50.0);
    }, /(Invalid period|Not enough valid data)/, 'Should throw error for insufficient data');
});

test('REVERSE_RSI nan handling', () => {


    const data = new Float64Array(50);
    for (let i = 0; i < 50; i++) {
        data[i] = i + 1.0;
    }

    const rsiLength = 14;
    const rsiLevel = 50.0;


    const dataWithNaN = new Float64Array(data);
    dataWithNaN[25] = NaN;

    const result = wasm.reverse_rsi_js(dataWithNaN, rsiLength, rsiLevel);
    assert.strictEqual(result.length, dataWithNaN.length);



    let hasValidAfter = false;
    for (let i = 26; i < result.length; i++) {
        if (!isNaN(result[i])) {
            hasValidAfter = true;
            break;
        }
    }
    assert(hasValidAfter, 'Should have valid values after warmup/NaN');


    const dataMultiNaN = new Float64Array(data);
    dataMultiNaN[20] = NaN;
    dataMultiNaN[30] = NaN;


    const result2 = wasm.reverse_rsi_js(dataMultiNaN, rsiLength, rsiLevel);
    assert.strictEqual(result2.length, dataMultiNaN.length);
    assert(result2 instanceof Float64Array, 'Should return Float64Array');
});

test('REVERSE_RSI warmup nans', () => {

    const close = new Float64Array(testData.close);
    const rsiLength = 14;
    const rsiLevel = 50.0;


    let firstValid = 0;
    for (let i = 0; i < close.length; i++) {
        if (!isNaN(close[i])) {
            firstValid = i;
            break;
        }
    }

    const result = wasm.reverse_rsi_js(close, rsiLength, rsiLevel);


    for (let i = 0; i < firstValid; i++) {
        assert(isNaN(result[i]), `Expected NaN at index ${i} (before first valid data)`);
    }
});

test('REVERSE_RSI memory management with into', () => {

    const close = new Float64Array(testData.close.slice(0, 50));
    const rsiLength = 14;
    const rsiLevel = 50.0;


    const outPtr = wasm.reverse_rsi_alloc(close.length);


    const inPtr = wasm.reverse_rsi_alloc(close.length);
    const wasmMemory = new Float64Array(wasm.__wasm.memory.buffer);
    const inOffset = inPtr / 8;


    for (let i = 0; i < close.length; i++) {
        wasmMemory[inOffset + i] = close[i];
    }


    wasm.reverse_rsi_into(inPtr, outPtr, close.length, rsiLength, rsiLevel);


    const outOffset = outPtr / 8;
    const output = new Float64Array(close.length);
    for (let i = 0; i < close.length; i++) {
        output[i] = wasmMemory[outOffset + i];
    }


    wasm.reverse_rsi_free(inPtr, close.length);
    wasm.reverse_rsi_free(outPtr, close.length);


    const expected = wasm.reverse_rsi_js(close, rsiLength, rsiLevel);
    assertArrayClose(output, expected, 1e-10, 'Memory-managed version should match normal version');
});

test('REVERSE_RSI batch processing', () => {

    const close = new Float64Array(testData.close.slice(0, 100));


    const config = {
        rsi_length_range: [10, 20, 5],
        rsi_level_range: [50.0, 50.0, 0.0]
    };

    const result = wasm.reverse_rsi_batch(close, config);


    assert(result.values, 'Should have values array');
    assert(result.combos, 'Should have combos array');
    assert.strictEqual(result.rows, 3, 'Should have 3 parameter combinations');
    assert.strictEqual(result.cols, close.length, 'Should have same columns as input');


    assert.strictEqual(result.combos.length, 3, 'Should have 3 combinations');
    const expectedLengths = [10, 15, 20];
    for (let i = 0; i < 3; i++) {
        assert.strictEqual(result.combos[i].rsi_length, expectedLengths[i],
            `Combo ${i} should have rsi_length ${expectedLengths[i]}`);
        assert.strictEqual(result.combos[i].rsi_level, 50.0,
            `Combo ${i} should have rsi_level 50.0`);
    }


    for (let row = 0; row < result.rows; row++) {
        const rowStart = row * result.cols;
        const rowData = result.values.slice(rowStart, rowStart + result.cols);


        let hasValid = false;
        for (let i = 0; i < rowData.length; i++) {
            if (!isNaN(rowData[i])) {
                hasValid = true;
                break;
            }
        }
        assert(hasValid, `Row ${row} should have some valid values`);
    }


    if (wasm.reverse_rsi_batch_into) {
        const len = close.length;


        const inPtr = wasm.reverse_rsi_alloc(len);
        const outPtr = wasm.reverse_rsi_alloc(len * 3);


        const wasmMemory = new Float64Array(wasm.__wasm.memory.buffer);
        const inOffset = inPtr / 8;
        for (let i = 0; i < len; i++) {
            wasmMemory[inOffset + i] = close[i];
        }


        const numRows = wasm.reverse_rsi_batch_into(
            inPtr, outPtr, len,
            10, 20, 5,
            50.0, 50.0, 0.0
        );

        assert.strictEqual(numRows, 3, 'Should return 3 rows');


        const outOffset = outPtr / 8;
        const output = new Float64Array(len * 3);
        for (let i = 0; i < output.length; i++) {
            output[i] = wasmMemory[outOffset + i];
        }


        wasm.reverse_rsi_free(inPtr, len);
        wasm.reverse_rsi_free(outPtr, len * 3);


        for (let row = 0; row < 3; row++) {
            const rowStart = row * len;
            const rowData = output.slice(rowStart, rowStart + len);

            let hasValid = false;
            for (let i = 0; i < rowData.length; i++) {
                if (!isNaN(rowData[i])) {
                    hasValid = true;
                    break;
                }
            }
            assert(hasValid, `Batch_into row ${row} should have some valid values`);
        }
    }
});

test('REVERSE_RSI numerical precision', () => {



    const extremeData = new Float64Array(40);
    for (let i = 0; i < 10; i++) {
        extremeData[i * 4] = 1e-10;
        extremeData[i * 4 + 1] = 1e10;
        extremeData[i * 4 + 2] = 1e-10;
        extremeData[i * 4 + 3] = 1e10;
    }

    const result1 = wasm.reverse_rsi_js(extremeData, 5, 50.0);
    assert.strictEqual(result1.length, extremeData.length);

    const validValues1 = result1.filter(v => !isNaN(v));
    assert(!validValues1.some(v => !isFinite(v)), 'Should not produce infinity');


    const smallDiffData = new Float64Array(50);
    for (let i = 0; i < 50; i++) {
        smallDiffData[i] = 100.0 + i * 1e-10;
    }

    const result2 = wasm.reverse_rsi_js(smallDiffData, 10, 50.0);
    assert.strictEqual(result2.length, smallDiffData.length);

    const validValues2 = result2.filter(v => !isNaN(v));
    if (validValues2.length > 0) {
        assert(!validValues2.some(v => !isFinite(v)), 'Should not produce infinity');
    }


    const constantData = new Float64Array(30);
    constantData.fill(100.0);

    const result3 = wasm.reverse_rsi_js(constantData, 10, 50.0);
    assert.strictEqual(result3.length, constantData.length);

    const validValues3 = result3.filter(v => !isNaN(v));
    if (validValues3.length > 0) {
        assert(!validValues3.some(v => !isFinite(v)), 'Should not produce infinity with constant values');
    }
});

File: tests/python/test_reverse_rsi.py
"""Python binding tests for REVERSE_RSI indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:

    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS
from rust_comparison import compare_with_rust


class TestReverseRsi:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()

    def test_reverse_rsi_partial_params(self, test_data):
        """Test REVERSE_RSI with partial parameters - mirrors check_reverse_rsi_partial_params"""
        close = test_data['close']


        result = ta_indicators.reverse_rsi(close, 14, 50.0, None)
        assert len(result) == len(close)

    def test_reverse_rsi_accuracy(self, test_data):
        """Test REVERSE_RSI accuracy - mirrors check_reverse_rsi_accuracy"""

        close = test_data['close']
        expected = EXPECTED_OUTPUTS['reverse_rsi']


        rsi_length = 14
        rsi_level = 50.0

        result = ta_indicators.reverse_rsi(close, rsi_length, rsi_level, None)


        assert len(result) == len(close)



        assert_close(
            result[-6:-1],
            expected['last_5_values'],
            rtol=1e-6,
            msg="REVERSE_RSI last 5 values mismatch"
        )

    def test_reverse_rsi_default_candles(self, test_data):
        """Test REVERSE_RSI with default parameters - mirrors check_reverse_rsi_default_candles"""
        close = test_data['close']


        result = ta_indicators.reverse_rsi(close, 14, 50.0, None)
        assert len(result) == len(close)

    def test_reverse_rsi_zero_period(self):
        """Test REVERSE_RSI fails with zero period - mirrors check_reverse_rsi_zero_period"""
        input_data = np.array([10.0, 20.0, 30.0])

        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.reverse_rsi(input_data, 0, 50.0, None)

    def test_reverse_rsi_period_exceeds_length(self):
        """Test REVERSE_RSI fails when period exceeds data length - mirrors check_reverse_rsi_period_exceeds_length"""
        data_small = np.array([10.0, 20.0, 30.0])

        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.reverse_rsi(data_small, 10, 50.0, None)

    def test_reverse_rsi_invalid_level(self):
        """Test REVERSE_RSI fails with invalid RSI level - mirrors check_reverse_rsi_invalid_level"""
        input_data = np.array([10.0, 20.0, 30.0, 40.0, 50.0] * 5)


        with pytest.raises(ValueError, match="Invalid RSI level"):
            ta_indicators.reverse_rsi(input_data, 14, 150.0, None)


        with pytest.raises(ValueError, match="Invalid RSI level"):
            ta_indicators.reverse_rsi(input_data, 14, -10.0, None)

    def test_reverse_rsi_edge_levels(self):
        """Test REVERSE_RSI with edge RSI levels (near 0 and 100) - not in Rust tests but validates edge behavior"""
        input_data = np.array([10.0, 20.0, 30.0, 40.0, 50.0] * 10)


        result = ta_indicators.reverse_rsi(input_data, 14, 0.01, None)
        assert len(result) == len(input_data)

        assert not np.all(np.isnan(result))


        result = ta_indicators.reverse_rsi(input_data, 14, 99.99, None)
        assert len(result) == len(input_data)

        assert not np.all(np.isnan(result))

    def test_reverse_rsi_various_levels(self):
        """Test REVERSE_RSI with various RSI levels"""
        input_data = np.array([10.0 + i for i in range(50)])

        levels = [20.0, 30.0, 50.0, 70.0, 80.0]
        results = []

        for level in levels:
            result = ta_indicators.reverse_rsi(input_data, 14, level, None)
            assert len(result) == len(input_data)
            results.append(result)


        for i in range(len(results) - 1):

            valid_mask = ~np.isnan(results[i]) & ~np.isnan(results[i+1])
            if np.any(valid_mask):
                assert not np.allclose(results[i][valid_mask], results[i+1][valid_mask], rtol=1e-10), \
                    f"Different RSI levels should produce different results"

    def test_reverse_rsi_empty_input(self):
        """Test REVERSE_RSI with empty input - mirrors check_reverse_rsi_empty_input"""
        input_data = np.array([])

        with pytest.raises(ValueError, match="Input data slice is empty"):
            ta_indicators.reverse_rsi(input_data, 14, 50.0, None)

    def test_reverse_rsi_all_nan(self):
        """Test REVERSE_RSI with all NaN values - mirrors check_reverse_rsi_all_nan"""
        input_data = np.array([np.nan] * 30)

        with pytest.raises(ValueError, match="All values are NaN"):
            ta_indicators.reverse_rsi(input_data, 14, 50.0, None)

    def test_reverse_rsi_insufficient_data(self):
        """Test REVERSE_RSI with insufficient data - mirrors check_reverse_rsi_very_small_dataset"""


        input_data = np.array([1.0, 2.0, 3.0, 4.0, 5.0])

        with pytest.raises(ValueError, match="(Invalid period|Not enough valid data)"):
            ta_indicators.reverse_rsi(input_data, 14, 50.0, None)

    def test_reverse_rsi_nan_handling(self):
        """Test REVERSE_RSI handles NaN in middle of data - mirrors check_reverse_rsi_nan_handling"""

        data = list(range(1, 51))
        input_data = np.array(data, dtype=np.float64)
        rsi_length = 14
        rsi_level = 50.0


        data_with_nan = input_data.copy()
        data_with_nan[25] = np.nan

        result = ta_indicators.reverse_rsi(data_with_nan, rsi_length, rsi_level, None)
        assert len(result) == len(data_with_nan)



        assert not np.all(np.isnan(result[26:])), "Should have valid values after warmup/NaN"


        data_multi_nan = input_data.copy()
        data_multi_nan[20] = np.nan
        data_multi_nan[30] = np.nan


        result2 = ta_indicators.reverse_rsi(data_multi_nan, rsi_length, rsi_level, None)
        assert len(result2) == len(data_multi_nan)
        assert isinstance(result2, np.ndarray)

    def test_reverse_rsi_warmup_nans(self, test_data):
        """Test REVERSE_RSI preserves warmup NaNs - mirrors check_reverse_rsi_warmup_nans"""
        close = test_data['close']
        rsi_length = 14
        rsi_level = 50.0


        first_valid = 0
        for i, val in enumerate(close):
            if not np.isnan(val):
                first_valid = i
                break

        result = ta_indicators.reverse_rsi(close, rsi_length, rsi_level, None)


        for i in range(first_valid):
            assert np.isnan(result[i]), f"Expected NaN at index {i} (before first valid data)"

    def test_reverse_rsi_stream(self, test_data):
        """Test REVERSE_RSI streaming functionality - mirrors check_reverse_rsi_streaming"""
        close = test_data['close'][:50]
        rsi_length = 14
        rsi_level = 50.0


        stream = ta_indicators.ReverseRsiStream(rsi_length, rsi_level)


        stream_results = []
        for value in close:
            result = stream.update(value)
            stream_results.append(result if result is not None else np.nan)


        batch_results = ta_indicators.reverse_rsi(close, rsi_length, rsi_level, None)



        stream_first_valid = next((i for i, v in enumerate(stream_results) if not np.isnan(v)), len(stream_results))
        batch_first_valid = next((i for i, v in enumerate(batch_results) if not np.isnan(v)), len(batch_results))


        first_valid = max(stream_first_valid, batch_first_valid)

        if first_valid < len(batch_results):

            assert_close(
                stream_results[first_valid:],
                batch_results[first_valid:],
                rtol=1e-3,
                msg="Stream and batch results mismatch"
            )

    def test_reverse_rsi_batch(self, test_data):
        """Test REVERSE_RSI batch processing with multiple RSI lengths - mirrors check_batch_sweep"""
        close = test_data['close'][:100]


        result = ta_indicators.reverse_rsi_batch(
            close,
            (10, 20, 5),
            (50.0, 50.0, 0),
            None
        )


        values = result['values']
        rsi_lengths_out = result['rsi_lengths']
        rsi_levels_out = result['rsi_levels']


        assert values.shape[0] == 3, "Should have 3 parameter combinations"
        assert values.shape[1] == len(close), "Should have same number of columns as input data"


        assert list(rsi_lengths_out) == [10, 15, 20], "RSI lengths should be [10, 15, 20]"
        assert np.allclose(rsi_levels_out, [50.0, 50.0, 50.0]), "RSI levels should all be 50.0"


        for i, length in enumerate([10, 15, 20]):
            row_data = values[i]
            assert len(row_data) == len(close)

            nan_count = np.sum(np.isnan(row_data))

            expected_warmup = (2 * length) - 1

            assert nan_count < len(row_data), f"Expected some non-NaN values after warmup with rsi_length {length}"

            valid_results = row_data[~np.isnan(row_data)]
            if len(valid_results) > 0:
                assert np.all(np.isfinite(valid_results)), "Reverse RSI values should be finite"

    def test_reverse_rsi_batch_different_levels(self, test_data):
        """Test REVERSE_RSI batch processing with different RSI levels"""
        close = test_data['close'][:100]


        result = ta_indicators.reverse_rsi_batch(
            close,
            (14, 14, 0),
            (30.0, 70.0, 20.0),
            None
        )


        values = result['values']
        rsi_lengths_out = result['rsi_lengths']
        rsi_levels_out = result['rsi_levels']


        assert values.shape[0] == 3, "Should have 3 parameter combinations"
        assert values.shape[1] == len(close), "Should have same number of columns as input data"


        assert list(rsi_lengths_out) == [14, 14, 14], "RSI lengths should all be 14"
        assert np.allclose(rsi_levels_out, [30.0, 50.0, 70.0]), "RSI levels should be [30.0, 50.0, 70.0]"


        for i in range(1, values.shape[0]):
            row_data = values[i]
            prev_row = values[i-1]
            valid_mask = ~np.isnan(row_data) & ~np.isnan(prev_row)
            if np.any(valid_mask):
                assert not np.allclose(row_data[valid_mask], prev_row[valid_mask], rtol=1e-10), \
                    f"Different RSI levels should produce different results"

    def test_reverse_rsi_kernel_consistency(self, test_data):
        """Test REVERSE_RSI produces consistent results across different kernels"""
        close = test_data['close'][:100]
        rsi_length = 14
        rsi_level = 50.0



        kernels = [
            (None, "Auto"),
            (1, "Scalar"),
            (2, "SSE2"),
        ]

        results = {}
        for kernel_value, kernel_name in kernels:
            try:
                result = ta_indicators.reverse_rsi(close, rsi_length, rsi_level, kernel_value)
                results[kernel_name] = result
            except Exception as e:

                print(f"Kernel {kernel_name} not available: {e}")
                continue


        if len(results) > 1:
            kernel_names = list(results.keys())
            base_kernel = kernel_names[0]
            base_result = results[base_kernel]

            for kernel_name in kernel_names[1:]:

                for i in range(len(base_result)):
                    if not np.isnan(base_result[i]) and not np.isnan(results[kernel_name][i]):
                        assert_close(
                            base_result[i],
                            results[kernel_name][i],
                            rtol=1e-12,
                            msg=f"Kernel {base_kernel} vs {kernel_name} mismatch at index {i}"
                        )
                    else:

                        assert np.isnan(base_result[i]) == np.isnan(results[kernel_name][i]), \
                            f"NaN mismatch between {base_kernel} and {kernel_name} at index {i}"

    def test_reverse_rsi_numerical_precision(self):
        """Test REVERSE_RSI numerical precision and edge cases"""

        extreme_data = np.array([1e-10, 1e10, 1e-10, 1e10] * 10, dtype=np.float64)
        result = ta_indicators.reverse_rsi(extreme_data, 5, 50.0, None)
        assert len(result) == len(extreme_data)

        assert not np.any(np.isinf(result[~np.isnan(result)])), "Should not produce infinity"


        small_diff_data = np.array([100.0 + i * 1e-10 for i in range(50)], dtype=np.float64)
        result = ta_indicators.reverse_rsi(small_diff_data, 10, 50.0, None)
        assert len(result) == len(small_diff_data)

        valid_values = result[~np.isnan(result)]
        if len(valid_values) > 0:
            assert not np.any(np.isinf(valid_values)), "Should not produce infinity"
            assert not np.any(np.isnan(valid_values)), "Valid values should not be NaN"


        constant_data = np.full(30, 100.0, dtype=np.float64)
        result = ta_indicators.reverse_rsi(constant_data, 10, 50.0, None)
        assert len(result) == len(constant_data)

        valid_values = result[~np.isnan(result)]
        if len(valid_values) > 0:
            assert not np.any(np.isinf(valid_values)), "Should not produce infinity with constant values"


if __name__ == "__main__":
    pytest.main([__file__, "-v"])